### DISCUSSION TITLE
Creating Custom Turtles

### DISCUSSION THREAD

**Discussion by aieis** [cc-tweaked/CC-Tweaked#1438] [Q&A] [answered]
Hey,

I am trying to wire a mod to cc tweaked. I have gauged that I need to create a turtle upgrade to use new items. I managed to get it working with a few vanilla items, but I've had no luck integrating it with the mod in question (TimelessandClassics_Reforged).

My intuition says that it has something to do with the item and the registry (perhaps I'm inappropriately handling deferred registration).

The mod compiles fine as well, it's just that nothing happens when I try using the modded tools instead of the vanilla ones with turtles.

The relevant pieces of code:
```
public class FirearmPeripheral extends AbstractTurtleUpgrade {

    protected FirearmPeripheral(ResourceLocation id, TurtleUpgradeType type, String adjective, NonNullSupplier<ItemStack> stack) {
        super(id, type, adjective, stack);
    }

    protected FirearmPeripheral(ResourceLocation id, TurtleUpgradeType type, NonNullSupplier<ItemStack> stack) {
        super(id, type, stack);
    } 
    .
    .
    .
    .
}
```
```
public class ComputerCraftRegister {
    public static void registerTurtleUpgrades(){
        ComputerCraftAPI.registerTurtleUpgrade(new FirearmPeripheral(new ResourceLocation( "minecraft", "test_equip" ),
                TurtleUpgradeType.BOTH, "test_equip", GLOCK_17::get));    }
}
```

As I mentioned, if I replace `GLOCK_17::get` with `Items.STICK` it works.

I am a programmer and consider myself experienced with Java (though I have been avoiding it more and more), but I am new to Minecraft and Minecraft modding. 

I have gone through the Minecraft Forge modding tutorial, but I have found that resources are quite scarce when it comes to mod/mod integrations. 

The details of my setup:
Minecraft 1.16.5
Forge 1.16.5-36.2.39-mdk
cc-tweaked-1.16.5-1.101.2 (recompiled locally)
tac-r-0.3.5.3-1.16.5 (and its required mods)

I would appreciate any help, general instructions, or resources I can use for reference.

**Comment by SquidDev** [comment]
How is `GLOCK_17` defined? That code looks fine to me.

**Reply by aieis** [reply]
The definition of the item in the source code of tac.

```
public static final RegistryObject<Item> GLOCK_17 = REGISTER.register("glock_17", () -> new TimelessPistolGunItem(properties -> properties.group(GunMod.PISTOL), GunModifiers.GLOCK17_MOD));
```

One more thing I just discovered. The turtle itself seems to exist, I managed to spawn a turtle with the correct model with the GLOCK_17 item attached.

My `getModel` function:
```
public TransformedModel getModel(@Nullable ITurtleAccess iTurtleAccess, @Nonnull TurtleSide side) {
        return TransformedModel.of( getCraftingItem(), side == TurtleSide.LEFT ? Transforms.leftTransform : Transforms.rightTransform );
}
```

Turtle still can't be crafted as usual.

Could the GLOCK_17 item itself be preventing the recipe from existing?

====================
References.

- definition of GLOCK_17
https://github.com/ClumsyAlien/TimelessandClassics_Reforged/blob/d0b9d87933f54abb31c77bac0144044d0430a724/src/main/java/com/tac/guns/init/ModItems.java#LL30C5-L30C5

- class TimelessPistolGunItem
https://github.com/ClumsyAlien/TimelessandClassics_Reforged/blob/d0b9d87933f54abb31c77bac0144044d0430a724/src/main/java/com/tac/guns/item/TransitionalTypes/TimelessPistolGunItem.java#L1

**Reply by SquidDev** [reply]
Very odd. Does your item have NBT at all on it? CC requires the NBT to match exactly, but you can change that behaviour.

**Reply by aieis** [reply]
Ah, that seems about right. I was able to craft the turtle with the item spawned through the game's console commands (which has different NBT data from the item taken from the creative menu or crafted properly).

It would be worth changing that behaviour, could you point me to that direction?

**Reply by SquidDev** [reply, marked as answer]
There's an [`isItemSuitable`](https://github.com/cc-tweaked/CC-Tweaked/blob/db2616d1c039b3a5fee1150bc2bfcdbc059e09be/src/main/java/dan200/computercraft/api/IUpgradeBase.java#L73-L85) method you can override, to ignore NBT. Note that the returned item (from `turtle.unequipNBT()` or whatever) will always have the NBT of the upgrade's item, not the item the turtle was crafted/equipped with.

**Reply by aieis** [reply]
Perfect. Thank you, it works as inteded for now. I'll stay vigilant to potential NBT-based bugs.
